File: redis_enterprise_admin.py
#!/usr/bin/env python3
import os
import sys
import json
import time
import requests
from requests.auth import HTTPBasicAuth
from urllib3.exceptions import InsecureRequestWarning
import urllib3

BASE_URL = os.getenv("REDIS_API_BASE_URL", "https://localhost:9443")
API_USER = os.getenv("REDIS_API_USER", "admin")
API_PASSWORD = os.getenv("REDIS_API_PASSWORD", "password")
DB_NAME = os.getenv("REDIS_DB_NAME", "python-demo-db")
TEMP_USER_PASSWORD = os.getenv("REDIS_TEMP_USER_PASSWORD", "ChangeMe123!")
VERIFY_TLS = os.getenv("REDIS_VERIFY_TLS", "false").lower() == "true"

if not VERIFY_TLS:
    urllib3.disable_warnings(InsecureRequestWarning)

session = requests.Session()
session.auth = HTTPBasicAuth(API_USER, API_PASSWORD)
session.headers.update({"Accept": "application/json"})
session.verify = VERIFY_TLS

def request(method, path, payload=None):
    url = f"{BASE_URL}{path}"
    headers = {}
    if payload is not None:
        headers["Content-Type"] = "application/json"

    resp = session.request(
        method=method,
        url=url,
        headers=headers,
        data=json.dumps(payload) if payload is not None else None,
        timeout=60,
    )
    if resp.status_code not in (200, 201, 204):
        raise RuntimeError(f"{method} {path} failed: HTTP {resp.status_code} -> {resp.text}")
    return resp

def create_database():
    # Redis Enterprise REST API expects the database configuration under the "bdb" field.
    # No modules are included here.
    payload = {
        "bdb": {
            "name": DB_NAME,
            "type": "redis",
            "memory_size": 1073741824,   # 1 GiB
            "shards_count": 1
        }
    }

    resp = request("POST", "/v1/bdbs", payload)
    data = resp.json()

    # Response may be the BDB object directly or nested in some deployments.
    db_obj = data if "uid" in data else data.get("bdb", {})
    db_uid = db_obj.get("uid")
    if db_uid is None:
        raise RuntimeError(f"Database create response did not contain uid: {data}")
    return db_uid

def create_user(name, email, role):
    payload = {
        "email": email,
        "password": TEMP_USER_PASSWORD,
        "name": name,
        "role": role,
        "email_alerts": False,
        "auth_method": "regular"
    }
    request("POST", "/v1/users", payload)
    print(f"Created user: {name} | {role} | {email}")

def list_users():
    resp = request("GET", "/v1/users")
    users = resp.json()

    print("\nUsers:")
    for u in users:
        name = u.get("name", "")
        role = u.get("role", "")
        email = u.get("email", "")
        print(f"{name} | {role} | {email}")

def delete_database(db_uid):
    request("DELETE", f"/v1/bdbs/{db_uid}")
    print(f"\nDeleted database uid={db_uid}")

def main():
    db_uid = create_database()
    print(f"Created database: {DB_NAME} (uid={db_uid})")

    create_user("John Doe", "john.doe@example.com", "db_viewer")
    create_user("Mike Smith", "mike.smith@example.com", "db_member")
    create_user("Cary Johnson", "cary.johnson@example.com", "admin")

    # Small wait can help in some lab environments if the user list is eventually consistent.
    time.sleep(1)
    list_users()

    delete_database(db_uid)

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

File: src/main/java/com/redis/assignment/exercise2/RedisEnterpriseAdmin.java
package com.redis.assignment.exercise2;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;
import java.time.Duration;
import java.util.Base64;

import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

public class RedisEnterpriseAdmin {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final String API_BASE_URL =
            getenv("REDIS_API_BASE_URL", "https://re-cluster1.ps-redislabs.org:9443");
    private static final String API_USER =
            getenv("REDIS_API_USER", "admin@rl.org");
    private static final String API_PASSWORD =
            getenv("REDIS_API_PASSWORD", "qST2pbF");
    private static final String DB_NAME =
            getenv("REDIS_DB_NAME", "new-db");
    private static final String TEMP_USER_PASSWORD =
            getenv("REDIS_TEMP_USER_PASSWORD", "ChangeMe123!");

    private static final long MEMORY_SIZE_BYTES = 1_073_741_824L; // 1 GiB
    private static final int SHARDS_COUNT = 1;

    private final HttpClient client;
    private final String basicAuthHeader;

    
    public RedisEnterpriseAdmin() throws Exception {
        this.client = buildUnsafeClient();
    
        String token = API_USER + ":" + API_PASSWORD;
        this.basicAuthHeader = "Basic " + Base64.getEncoder()
                .encodeToString(token.getBytes(StandardCharsets.UTF_8));
    }

    private static HttpClient buildUnsafeClient() throws Exception {
        TrustManager[] trustAllCerts = new TrustManager[] {
                new X509TrustManager() {
                    @Override
                    public void checkClientTrusted(X509Certificate[] chain, String authType) { }
    
                    @Override
                    public void checkServerTrusted(X509Certificate[] chain, String authType) { }
    
                    @Override
                    public X509Certificate[] getAcceptedIssuers() {
                        return new X509Certificate[0];
                    }
                }
        };

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, trustAllCerts, new SecureRandom());
    
        return HttpClient.newBuilder()
                .sslContext(sslContext)
                .connectTimeout(Duration.ofSeconds(15))
                .build();
    }
    

    public static void main(String[] args) throws Exception {
        new RedisEnterpriseAdmin().run();
    }

    private void run() throws Exception {
        long dbUid = createDatabase();
        System.out.println("Created database uid=" + dbUid + " name=" + DB_NAME);

        //as the cluster is RBAC-enabled, we'll use role_uids instead of role in the request body
        //and because initialized db only has role 'admin' with uid = 1
        //we need to create two roles required in this exercises (db_viewer and db_member)
        //curl -X POST -H "accept: application/json" -H "Content-Type: application/json" -u "admin@rl.org:qST2pbF" https://re-cluster1.ps-redislabs.org:9443/v1/roles -d '{"name":"db_viewer", "management": "db_viewer"}' -k -i
        //curl -X POST -H "accept: application/json" -H "Content-Type: application/json" -u "admin@rl.org:qST2pbF" https://re-cluster1.ps-redislabs.org:9443/v1/roles -d '{"name":"db_member", "management": "db_member"}' -k -i

        //lines for user creation below may fail if this java code has been executed before (hence same emails already registered for existing users)
        //you may delete these users first, easier using curl like below (assume uids are 2,3,4)
        //curl -X DELETE -H "accept: application/json" -u "admin@rl.org:qST2pbF" https://re-cluster1.ps-redislabs.org:9443/v1/users/2 -k -i
        //curl -X DELETE -H "accept: application/json" -u "admin@rl.org:qST2pbF" https://re-cluster1.ps-redislabs.org:9443/v1/users/3 -k -i
        //curl -X DELETE -H "accept: application/json" -u "admin@rl.org:qST2pbF" https://re-cluster1.ps-redislabs.org:9443/v1/users/4 -k -i
        try {
            createUser("John Doe", "john.doe@example.com", 4);
            createUser("Mike Smith", "mike.smith@example.com", 5);
            createUser("Cary Johnson", "cary.johnson@example.com", 1);
        }catch(Exception e){
            System.out.println("Exception on create user: " + e.getMessage());    
        }
        System.out.println("\nUsers:");
        listAndDisplayUsers();

        deleteDatabaseWithRetry(dbUid);
        System.out.println("\nDeleted database uid=" + dbUid);
    }

    private long createDatabase() throws Exception {
        ObjectNode bdb = MAPPER.createObjectNode();
        bdb.put("name", DB_NAME);
        bdb.put("type", "redis");
        bdb.put("memory_size", MEMORY_SIZE_BYTES);
        bdb.put("shards_count", SHARDS_COUNT);

        ObjectNode body = MAPPER.createObjectNode();
        body.set("bdb", bdb);

        HttpResponse<String> resp = sendJson("POST", "/v1/bdbs", bdb);
        ensureSuccess(resp, 200, 201);

        JsonNode json = parseJson(resp.body());

        // Create responses may return the BDB object directly or nested under "bdb".
        JsonNode dbNode = json.has("uid") ? json : json.path("bdb");
        long uid = dbNode.path("uid").asLong(-1);
        if (uid < 0) {
            throw new IllegalStateException("Create database response did not contain uid: " + resp.body());
        }
        return uid;
    }

    private void createUser(String name, String email, int roleId) throws Exception {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("email", email);
        body.put("password", TEMP_USER_PASSWORD);
        body.put("name", name);
        body.put("email_alerts", false);
        body.put("auth_method", "regular");

        ArrayNode roleUids = body.putArray("role_uids");
        roleUids.add(roleId);

        HttpResponse<String> resp = sendJson("POST", "/v1/users", body);
        ensureSuccess(resp, 200, 201);

        System.out.printf("Created user: %s (%s)%n", name, email);
    }

    private void listAndDisplayUsers() throws Exception {
        HttpResponse<String> resp = sendJson("GET", "/v1/users", null);
        ensureSuccess(resp, 200);

        JsonNode users = parseJson(resp.body());
        if (!users.isArray()) {
            throw new IllegalStateException("Expected an array from GET /v1/users, got: " + resp.body());
        }

        for (JsonNode user : users) {
            String name = user.path("name").asText("");
            String role = user.hasNonNull("role")
                    ? user.get("role").asText("")
                    : user.path("role_uids").toString();
            String email = user.path("email").asText("");

            System.out.printf("%s | %s | %s%n", name, role, email);
        }
    }

    private void deleteDatabaseWithRetry(long uid) throws Exception {
        long deadlineMs = System.currentTimeMillis() + 60_000L;

        while (System.currentTimeMillis() < deadlineMs) {
            HttpResponse<String> resp = sendJson("DELETE", "/v1/bdbs/" + uid, null);

            if (resp.statusCode() == 200 || resp.statusCode() == 204) {
                return;
            }

            // Database may still be becoming active; Redis can return 409 while busy or not ready.
            if (resp.statusCode() == 409) {
                Thread.sleep(2000);
                continue;
            }

            throw new IllegalStateException(
                    "Failed to delete database uid=" + uid + " status=" + resp.statusCode() + " body=" + resp.body());
        }

        throw new IllegalStateException("Timed out waiting to delete database uid=" + uid);
    }

    private HttpResponse<String> sendJson(String method, String path, JsonNode body) throws Exception {
        HttpRequest.Builder builder = HttpRequest.newBuilder()
                .uri(URI.create(API_BASE_URL + path))
                .timeout(Duration.ofSeconds(60))
                .header("Authorization", basicAuthHeader)
                .header("Accept", "application/json");

        if (body != null) {
            builder.header("Content-Type", "application/json");
        }

        HttpRequest request;
        switch (method) {
            case "GET":
                request = builder.GET().build();
                break;
            case "POST":
                request = builder.POST(HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body))).build();
                break;
            case "DELETE":
                request = builder.DELETE().build();
                break;
            default:
                throw new IllegalArgumentException("Unsupported method: " + method);
        }

        return client.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private static void ensureSuccess(HttpResponse<String> resp, int... allowed) {
        for (int code : allowed) {
            if (resp.statusCode() == code) {
                return;
            }
        }
        throw new IllegalStateException("HTTP " + resp.statusCode() + ": " + resp.body());
    }

    private static JsonNode parseJson(String body) throws Exception {
        return MAPPER.readTree(body);
    }

    private static String getenv(String key, String defaultValue) {
        String value = System.getenv(key);
        return (value == null || value.isBlank()) ? defaultValue : value;
    }
}